### ISSUE TITLE
Big method call overhead (up to 4x) due to unnecessary vector allocations.

### ISSUE THREAD

**Issue by Josef-Vonasek** [jni-rs/jni-rs#252]
I benchmarked a code allocates 2 million java objects, with `env.new_method_unchecked`. 

To my surprise, the code took ~1 second to execute, or ~1ms/allocation. 
This is ~10x slower than what I would get if I initialized those objects in java directly.

So I decided to profile the application and got the following stacktrace:

![image](https://user-images.githubusercontent.com/22682767/93091048-e75ab000-f69d-11ea-9312-59c1e13c9ade.png)

My objects take exactly one integer argument, and as you can see, the vector allocations are responsible for 80% of time spend in `env.new_object_unchecked`.

In other words, the method call could be 4x faster, if it didn't allocate this vector on each call and used the vararg `env.NewObjectV` under the hood.

**Comment by dmitry-timofeev** [comment]
Thanks for the report! The throughput improvement here would certainly be welcome.

Part of the difference might come from using different methods: `NewObject` in C and `NewObjectA` [in Rust](https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/jnienv.rs#L940). Here is a specification of the differences: https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#newobject-newobjecta-newobjectv. It is possible to double-check if that has any effect by modifying your C benchmark to use `NewObjectA`.

The second difference is that the C benchmark doesn't do what it should according to the JNI spec: check for possible uncaught exceptions thrown in Java code. Even `...unchecked` family of methods in Rust do that (with some exceptions for cases when it is not required by the JNI spec): https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/macros.rs#L21 

As that is a spec requirement, I'd also suggest to add that to the C benchmark so that we eventually get to see pure Rust overhead.

**Comment by dmitry-timofeev** [comment]
Also, some benchmarks we have might help: https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#run-benchmarks

**Comment by Josef-Vonasek** [comment]
I have crated a [repo with benchmarks](https://github.com/enso-org/jvm-native-benchmarks) that compare C++, jni-rs and unsafe rust.

**Comment by dmitry-timofeev** [comment]
:+1: Have you managed to confirm that the difference comes from using `NewObjectA` (not mainly from checking for exceptions, which is required by the spec and which we cannot change)?
